Allow using - to indicate input from stdin (resolves #305)

NieDzejkob opened the pull request.

NieDzejkob commented:
Sorry, I should've looked at `CONTRIBUTING.rst` :sweat_smile:

AntonioND commented:
Hmm, this is weird. I'm testing with this test file (``a.asm``):
```
SECTION "a",ROM0
DB 0, 1, 2, 3
```
If I do this, it fails:
```
$ cat a.asm | rgbasm -o a.obj -
malloc(): corrupted top size
```
While if I do this, it works:
```
$ rgbasm -o a.out - < a.asm
```
Does it happen to you?

Ben10do commented:
Just to come out of the woodwork for once…

Running all three examples with Clang's AddressSanitizer (i.e. compiling with `-fsanitize=address`) produced the same effects: only the version using `cat` failed. (Worryingly, it worked on my machine when I *didn't* AddressSan 😟)

I think it's because, in `lexer.c:161`, we try to seek to the end of the file, to determine how large it is (and is used to create a buffer). This doesn't make sense when using piping, because the other program may still be running and producing output.

Off the top of my head, I'm not sure what the best way of getting around this is, since we do a fair bit of preprocessing using said buffer. Maybe the only solution is to wait until the process providing stdin has finished?

Full output of AddressSan:
```
==99165==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x619000001d82 at pc 0x00010b190aa4 bp 0x7ffee4c25060 sp 0x7ffee4c24808
WRITE of size 31 at 0x619000001d82 thread T0
    #0 0x10b190aa3 in wrap_fread (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x1eaa3)
    #1 0x10aff2d01 in yy_create_buffer lexer.c:176
    #2 0x10afefa52 in fstk_Init fstack.c:427
    #3 0x10afff08d in main main.c:439
    #4 0x7fff71e94ed8 in start (libdyld.dylib:x86_64+0x16ed8)

0x619000001d82 is located 0 bytes to the right of 1026-byte region [0x619000001980,0x619000001d82)
allocated by thread T0 here:
    #0 0x10b1c8f53 in wrap_malloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56f53)
    #1 0x10aff2b0a in yy_create_buffer lexer.c:168
    #2 0x10afefa52 in fstk_Init fstack.c:427
    #3 0x10afff08d in main main.c:439
    #4 0x7fff71e94ed8 in start (libdyld.dylib:x86_64+0x16ed8)

SUMMARY: AddressSanitizer: heap-buffer-overflow (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x1eaa3) in wrap_fread
Shadow bytes around the buggy address:
  0x1c3200000360: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200000370: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200000380: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200000390: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c32000003a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x1c32000003b0:[02]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c32000003c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c32000003d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c32000003e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c32000003f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3200000400: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==99165==ABORTING
fish: Process 99165, './rgbasm' 'cat a.asm | ./rgbasm -o a.obj -' terminated by signal SIGABRT (Abort)
```

NieDzejkob commented:
Hmm, I certainly didn't expect this change to affect such internals. In this situation, I am considering modifying rgbasm to only perform a single pass over the source code. This shouldn't be a significant change in the complexity, but since it's not a minor change, I would like to first request a comment on this matter from people more experienced with the codebase.

The basic idea is to write out generated machine code into memory in the first pass, leaving holes as if the referenced symbols are external. Then, when a label is defined, one can simply change the symbol into an export or local, and resolve all patches generated so far.

I would expect this to also be beneficial in terms of performance of the program.

AntonioND commented:
I'm not too sure that you can remove the second pass of the assembler... Even if you do, what are you going to do about the assumptions the code makes about reading the whole file in one go? I think that's a bigger problem.

dbrotz commented:
Fixing this properly would require rewriting most of the lexer and macro code, at the least. Removing the second pass would pretty much require rewriting the whole program.

However, it would be pretty simple to just buffer all the input from stdin until the whole file is read (and leave the rest of the program unchanged). It would be slower than the ideal implementation since it would involve additional copying of memory.

AntonioND commented:
Yes, I agree that's not viable without a lot of effort. At that point it is better to rewrite rbgasm completely.

And I agree, your proposal would solve the problem. I wouldn't be too concerned about performance or extra memory used, though.

dbrotz commented:
I may have spoken too soon regarding the removal of the second pass. I was able to remove it without too much trouble since this assembler defers so much to the linker. The only downside seems to be that errors for undefined labels can only be caught at the end.

AntonioND commented:
I guess that now that the second pass has been removed this could work.

NieDzejkob commented:
I've rebased the patch onto current master and fixed the issue with using actual pipes as input. One of the tests I did was building pokecrystal with the Makefile modified so that the input is passed with `cat.

I added an automated test for the feature, but I'm not sure it's the best possible way to test it. Let me know if you have any better ideas.

mid-kid commented:
I wouldn't make every test run in pipe mode as well, that just increases the amount of test files for little purpose. I'd just build one of the most basic tests in pipe mode - if that works, everything should.

mid-kid commented:
Also, it seems like your patch will handle the case where `INCLUDE "-"` is used in a source file. I don't think that's desireable. Coud that be improved such that "-" will only be interpreted as stdin whe supplied from the command-line?

mid-kid commented:
Last thing, could you handle this the same for `-o` where "-" indicates wiriting to stdout? Or would that be too difficult/should be handled in a different PR.

Ben10do commented on src/asm/lexer.c:
```diff
@@ -156,25 +156,39 @@ YY_BUFFER_STATE yy_create_buffer(FILE *f)
-	pBuffer->pBufferStart = pBuffer->pBufferRealStart + SAFETYMARGIN;
-	pBuffer->pBuffer = pBuffer->pBufferRealStart + SAFETYMARGIN;
+		char *bufpos = buf + SAFETYMARGIN + size;
+		size_t read_count = fread(bufpos, 1, capacity - size, f);
```
While I don't think it makes any functional difference, I think we keep the `sizeof(uint8_t)` in place, rather than replacing it with the hardcoded 1, to make it clearer what that parameter represents.

Ben10do commented:
In response to @mid-kid's comments:

Firstly, I think your concerns about the tests are valid, and I've touched on the tradeoff in my last comment.

Secondly, this PR doesn't appear to allow `INCLUDE "-"`, as the check for `-` is only done in `fstk_Init()`, which is only run for the main file (but thanks for prompting me to check!).

Thirdly, although that sounds like a natural counterpart, I think it's better to leave it for a later PR 🙂

mid-kid commented:
>Secondly, this PR doesn't appear to allow INCLUDE "-", as the check for - is only done in fstk_Init(), which is only run for the main file (but thanks for prompting me to check!).

Ah, sorry for the misunderstanding. I was under the impression that this fstk function was called for every file it tried to open and parse.

NieDzejkob commented on src/asm/lexer.c:
```diff
@@ -156,25 +156,39 @@ YY_BUFFER_STATE yy_create_buffer(FILE *f)
 
-	/* Give extra room for 2 newlines and terminator */
-	uint32_t capacity = size + 3;
+	if (capacity == -1)
```
`ftell` will not return EOF. If the end of file has been reached, the size of the file will be returned. `ftell` is specified to return `-1` in case of error, see `ftell(3)` or `ftell(3p)`

Ben10do replied:
Ah, apologies, it seems like I was looking at [duff information on cppreference.com](https://en.cppreference.com/mwiki/index.php?title=c/io/ftell&oldid=78945) (though, strangely, [the corresponding C++ page gets it right](https://en.cppreference.com/w/cpp/io/c/ftell)).

Either way, thanks for looking into resolving the `capacity == 0` issue! It probably doesn't need to be a separate condition to the above, but whatever 🤷‍♂️

NieDzejkob commented on src/asm/lexer.c:
```diff
@@ -156,25 +156,39 @@ YY_BUFFER_STATE yy_create_buffer(FILE *f)
-	size = ftell(f);
-	fseek(f, 0, SEEK_SET);
+	capacity = ftell(f);
+	rewind(f);
```
The logic here is that on a pipe, all three calls will fail, giving `capacity = -1`. After reading some of the POSIX spec for these functions, I can see a better implementation, which I will push in a moment---"The behavior of fseek() on devices which are incapable of seeking is implementation-defined.  The value of the file offset associated with such a device is undefined."

Ben10do replied:
Thanks for updating this; the new behaviour seems reasonable to me.

NieDzejkob commented:
I rebased onto current master and modified the code responsible for determining the initial buffer capacity after some comments from Ben.

NieDzejkob commented:
Pushed again because of a screwed-up attempt to please checkpatch.

Ben10do commented on src/asm/lexer.c:
```diff
@@ -156,28 +156,71 @@ YY_BUFFER_STATE yy_create_buffer(FILE *f)
+	/* This is added here to make the buffer scaling above easy to express,
+	 * while taking the newline space into account
+	 * for the `yy_buffer_append`s below.
+	 */
```
I think you've tried to make the old "Give extra room for 2 newlines and terminator" comment a little clearer, while also taking away the reasoning for this magic number… 🙂 If you're happy to reword this, to make it clear here that the later `yy_buffer_append`s might add two newlines and a terminator (justifying the `3` here), that'd be great!

AntonioND commented:
This new behavior needs to be documented.
